mesa/mesaPROTON_OC.py

```diff
@@ -367,30 +368,38 @@ def total_num_links(self):
             for a in self.schedule.agents]) / 2
 
     def setup_oc_groups(self):
+        """
+        This procedure creates "criminal" type links within the families, in case there are not enough members
+        takes members from outside.
+        :return: None
+        """
         # OC members are scaled down if we don't have 10K agents
-        scaled_num_oc_families = math.ceil(
-            self.num_oc_families * self.initial_agents / 10000 * self.num_oc_persons / 30)
-        scaled_num_oc_persons = math.ceil(self.num_oc_persons * self.initial_agents / 10000)
-        # families first. Note that it could extract the same family twice. This could be improved to force exactly the number of families needed.
+        scaled_num_oc_families = np.ceil(self.num_oc_families * self.initial_agents / 10000 * self.num_oc_persons / 30)
+        scaled_num_oc_persons = np.ceil(self.num_oc_persons * self.initial_agents / 10000)
+        # families first.
         # we assume here that we'll never get a negative criminal tendency.
-        oc_family_head = extra.weighted_n_of(scaled_num_oc_families,
-                                             self.schedule.agents, lambda x: x.criminal_tendency, self.rng)
-        for x in oc_family_head: x.oc_member = True
-        candidates_in_families = [y for y in x.neighbors.get('household') if y.age() >= 18 for x in oc_family_head]
-        if len(candidates_in_families) >= scaled_num_oc_persons - scaled_num_oc_families:  # family members will be enough
-            members_in_families = extra.weighted_n_of(scaled_num_oc_persons - scaled_num_oc_families,
-                                                      candidates_in_families, lambda x: x.criminal_tendency, self.rng)
+        oc_family_heads = extra.weighted_n_of(scaled_num_oc_families, self.schedule.agents, lambda x: x.criminal_tendency, self.rng)
+        candidates = list()
```

Comment:
Why?

Reply:
Referring to the [Netlogo code]( https://github.com/LABSS/PROTON-OC/blob/master/PROTON-OC.nlogo#L679-L687) , I followed the same procedure. The line that here indicates as deleted is actually moved to line 387.

Reply:
No, that was why the candidate=list(). Wouldn't it work also without?

Why you changed the select procedure to a for cycle? Didn't ` candidates_in_families = [y for y in x.neighbors.get('household') if y.age() >= 18 for x in oc_family_head]` work as well?

Reply:
Okay, I get it! I did not use that list comprehension because it is incorrect! In any case I did a speed-test between the correct version of the  list comprehension and the currently used method. I share the results because they could be useful in the future. Of course I will implement the fastest version!

![image](https://user-images.githubusercontent.com/3855193/97578446-17150b80-19f1-11eb-8841-d35c4e9fc0f0.png)

Result:

![image](https://user-images.githubusercontent.com/3855193/97578549-3c097e80-19f1-11eb-8dbe-dafcff02112b.png)